---
title: "phyloseq_4_Functions for Accessing and (Pre)Processing Data"
author: "wentao"
date: "2019/8/23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```




## 访问的预处理数据
**Functions for Accessing and (Pre)Processing Data**
这篇教程于2018年三月更新

为了保证可重复性，这里显示的phyloseq的版本号。

><font size=2>For completeness, here is the version number of phyloseq used to build this instance of the tutorial – and also how you can check your own current version from the command line.

## 数据分析之前
被微生物组数据虐过的生信工作者都知道，其实流程化的做完整套分析是容易的；最为耗费心力的就是在数据不好的情况下，对微生物组数据的过滤。
通过本套教程，我们将学会如何使用分类信息过滤数据，如何过滤低丰度OTU，如何在不同分类等级合并otu丰度信息。



### 数据分析之前：数据访问
**Accessors**
phyloseq包含我们进行分析的五个部分文件，比如OTU表格，可以通过特殊的访问函数进行访问，这些访问内容的函数可以共用户进行各种分析或者与phyloseqa之外的函数进行交互。


><font size=2>Components of a phyloseq object, like the OTU Table, can be accessed by special accessor functions, or ``accessors’’, which return specific information about phylogenetic sequencing data, if present. These accessor functions are available for direct interaction by users and dependent functions/packages.


```{R}
library(phyloseq)
packageVersion("phyloseq")
data("GlobalPatterns")

GlobalPatterns
```

### 对群落数据进行粗略统计和分析

#### 统计OTU数量
```{R}
ntaxa(GlobalPatterns)
```
#### 统计样品数量
```{R}
nsamples(GlobalPatterns)
```

#### 查看样品名称
```{R}

sample_names(GlobalPatterns)[1:5]
```
#### 查看物种分类等级
```{R}

rank_names(GlobalPatterns)
```
#### 查看分组文件表头信息
```{R}
sample_variables(GlobalPatterns)

```
#### 部分可视化OTU表格
```{R}
otu_table(GlobalPatterns)[1:5, 1:5]

```
#### 部分可视化tax注释表格
```{R}

tax_table(GlobalPatterns)[1:5, 1:4]
```
#### 查看进化树文件

```{R}
phy_tree(GlobalPatterns)

```

```{R}
taxa_names(GlobalPatterns)[1:10]

```

### 提取丰度排名前十位的的OTU，初步使用进化树可视化

提取子集，可视化进化树
```{R}
myTaxa = names(sort(taxa_sums(GlobalPatterns), decreasing = TRUE)[1:10])
ex1 = prune_taxa(myTaxa, GlobalPatterns)
plot(phy_tree(ex1), show.node.label = TRUE)

```

之前大家都知道了作者想突出phyloseq进化树的可视化方案。那么plot_tree到底有什么特色呢？
下面随着作者的脚部，我们一点一点揭开。
plot_tree对叶节点的修饰工作是目前我见过最为丰度和简单的：

- 按照门水平做进化树
- 叶节点添加一组圆点，大小用于展示不同分组门水平物种丰度。

以后在进化树那一章教程中我们详细解读作者的plot_tree


```{R}
plot_tree(ex1, color = "SampleType", label.tips = "Phylum", ladderize = "left", justify = "left" , size = "Abundance")

```

### 数据预处理
**Preprocessing**
phyloseq包含有函数用于数据过滤，取子集，合并OTU表格文件等，许多函数借鉴了genefilter包。prune_taxa和prune_samples函数直接去除不想要的OTU和样品。filterfun_sample和genefilter_sample用于复杂的过滤条件设计并过滤。在下面的案例中GlobalPatterns数据集首先转化为相对丰度，然后构造新的GPr对象，过滤丰度在10^-5一下的OTU。

译者补充：使用相对丰度标准化OTU表格并过滤OTU

><font size=2>The phyloseq package also includes functions for filtering, subsetting, and merging abundance data. Filtering in phyloseq is designed in a modular fashion similar to the approach in the genefilter package. This includes the prune_taxa and prune_samples methods for directly removing unwanted indices, as well as the filterfun_sample and genefilter_sample functions for building arbitrarily complex sample-wise filtering criteria, and the filter_taxa function for taxa-wise filtering. In the following example, the GlobalPatterns data is first transformed to relative abundance, creating the new GPr object, which is then filtered such that only OTUs with a mean greater than 10^-5 are kept.

```{R}

GPr  = transform_sample_counts(GlobalPatterns, function(x) x / sum(x) )
GPfr = filter_taxa(GPr, function(x) mean(x) > 1e-5, TRUE)
GPr
GPfr
```
以上操作提取了数据子集GPfr，这一子集包含4624个OTU，而源数据包含大约19216个OTU。以上两行代码我们通过构造了自定义过滤函数并使用该函数进行过滤。

><font size=2>This results in a highly-subsetted object, GPfr, containing just 4624 of the original ~19216 OTUs. Note that in both lines we have provided a custom function for transformation and filtering, respectively.

### subset_taxa：根据分类信息提取部分OTU;prune_samples：根据丰度过滤样品
对于我们想要的样品和OTU可以通过prune_taxa和prune_samples函数进行提取。这两个函数可以整合tax注释表格信息，也就是说可以通过注释表格对OTU进行提取，而phyloseq封装体系会整个变化。下面的函数提取门为Chlamydiae的OTU，下一条代码为提取样品序列总的数量超过20条的样品。


><font size=2>The subsetting methods prune_taxa and prune_samples are for cases where the complete subset of desired OTUs or samples is directly available. Alternatively, the subset_taxa and subset_samples functions are for subsetting based on auxiliary data contained in the Taxonomy Table or Sample Data components, respectively. These functions are analogous to the subset function in core R, in which the initial data argument is followed by an arbitrary logical expression that indicates elements or rows to keep. Thus, entire experiment-level data objects can be subset according to conditional expressions regarding the auxiliary data. For example, the following code will first assign to GP.chl the subset of the GlobalPatterns dataset that are part of the Chlamydiae phylum, and then remove samples with less than 20 total reads.

```{R}
?subset_taxa
GP.chl = subset_taxa(GlobalPatterns, Phylum=="Chlamydiae")
GP.chl
GP.chl = prune_samples(sample_sums(GP.chl)>=20, GP.chl)
GP.chl
```
### 复习一下merge_taxa函数，用于合并指定的OTU

合并函数用于合并指定的OTU，这些函数有用于OTU合并和样品合并的函数：merge_taxa和merge_samples。还有merge_phyloseq用于合并phyloseq中的不同对象。以下代码合并仅在Chlamydiae数据集中的前5个OTU。

><font size=2>Merging methods include merge_taxa and merge_samples, intended for merging specific OTUs or samples, respectively. There is also the merge_phyloseq function for a complete merge of two or more phyloseq-objects (or a phyloseq-object and one or more separate components). For example, the following code merges the first 5 OTUs in the Chlamydiae-only dataset.

```{R}
GP.chl.merged = merge_taxa(GP.chl, taxa_names(GP.chl)[1:5])
GP.chl.merged 
```
### tax_glom函数：根据注释等级合并OTU

在merge_taxa函数的基础上，phyloseq包包含量两个升级函数：tip_glom和tax_glom；通过进化树和物种分类等级进行合并。下面的例子我们通过科水平合并数据并进行可视化。

><font size=2>Building on the merge_taxa methods, the phyloseq-package includes the agglomeration functions, tip_glom and tax_glom, for merging all OTUs in an experiment that are similar beyond a phylogenetic or taxonomic threshold, respectively. The following code demonstrates how to agglomerate the “Bacteroidetes-only” dataset (called gpsfb) at the taxonomic rank of Family, and create an annotated tree of the result.

未发现gpsfb数据集,这里我们使用GP.chl数据集替代。 
```{R}
gpsfb = GP.chl

gpsfbg = tax_glom(gpsfb, "Family")
gpsfbg
plot_tree(gpsfbg, color="SampleType", shape="Class", size="abundance")


```
### otu表格进行相对丰度标准化。

transform_sample_counts函数是phyloseq包中提供的对数据进行转化的典型函数。通过自定义函数对数据进行高度自定义的转化，下面的案例我们将OTU进行相对丰度转化。

><font size=2>For transforming abundance values by an arbitrary R function, phyloseq includes the transform_sample_counts function. It takes as arguments a phyloseq-object and an R function, and returns a phyloseq-object in which the abundance values have been transformed, sample-wise, according to the transformations specified by the function. For example, the following command transforms GP.chl abundance counts to fractional abundance.

```{R}
transform_sample_counts(GP.chl, function(OTU) OTU/sum(OTU) )

```

### filter_taxa：对otu表格的过滤
同subset_taxa函数不同,filter_taxa函数可以写函数对otu进行过滤；下面例子为保留在超过20%样品中的超过3条序列的OTU。

><font size=2>Finally, the following is the remaining set of preprocessing steps that was applied to the GlobalPatterns OTU counts prior to creating the figures in the main phyloseq manuscript.

><font size=2>Remove taxa not seen more than 3 times in at least 20% of the samples. This protects against an OTU with small mean & trivially large C.V.

```{R}
GP = filter_taxa(GlobalPatterns, function(x) sum(x > 3) > (0.2*length(x)), TRUE)

```
### mapping文件修改
此处添加一列

```{R}
sample_data(GP)$human = factor( get_variable(GP, "SampleType") %in% c("Feces", "Mock", "Skin", "Tongue") )

```

### 抽平
这里使用中位数抽平
```{R}
total = median(sample_sums(GP))
standf = function(x, t=total) round(t * (x / sum(x)))
gps = transform_sample_counts(GP, standf)

```

### 这里提取变异系数大的OTU，为了放大差异
gpss数据集未找到,所以后面几条命令无法运行,所以我们使用GP数据集做一个替代。


```{R}

gpss = GP
gpsf = filter_taxa(gps, function(x) sd(x)/mean(x) > 3.0, TRUE)
gpsf
```
### 根据门类进行提取子集

```{R}

gpsfb = subset_taxa(gpsf, Phylum=="Bacteroidetes")

```
### 柱状图展示丰度信息
```{R}
title = "plot_bar; Bacteroidetes-only"
plot_bar(gpsfb, "SampleType", "Abundance", title=title)

```
### 展示科水平丰度信息
```{R}
plot_bar(gpsfb, "SampleType", "Abundance", "Family", title=title)

```
### 这里展示科水平丰度信息，使用分面展示不同的科
```{R}
plot_bar(gpsfb, "Family", "Abundance", "Family", 
         title=title, facet_grid="SampleType~.")

```

### reference
https://joey711.github.io/phyloseq/preprocess.html


