---
title: "phyloseq_H4_Powerful, flexible phyloseq bar plots"
author: "wentao"
date: "2019/8/23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 强大而又灵活的phyloseq柱状图系统
**Powerful, flexible phyloseq bar plots**

下面的例子将叫你如何使用自己的数据做柱状图。

><font size=2>The following are examples to help get you started using the plot_bar function on your own phyloseq data

## 本次使用的还是GlobalPatterns数据集
**Global Patterns dataset examples**

```{R}
library("phyloseq"); packageVersion("phyloseq")
library("ggplot2"); packageVersion("ggplot2")

theme_set(theme_bw())


```



译者补充：subset_taxa函数，提取子集，这里提取门为Chlamydiae的OTU。

```{R}

data("GlobalPatterns")
gp.ch = subset_taxa(GlobalPatterns, Phylum == "Chlamydiae")
```


### 一些初步的柱状图
**Some initial basic plots**

以下是未调整参数的默认条形图。默认情况下，每个样本映射到X轴坐标上，丰度信息映射到Y轴上。OTU排布信息按照丰度进行排序。只要你选择展示的OTU不止一个，就会根据OTU在样本中的平均丰度进行排序。

><font size=2>The following is the default barplot when no parameters are given. The dataset is plotted with every sample mapped individually to the horizontal (x) axis, and abundance values mapped to the veritcal (y) axis. At each sample’s horizontal position, the abundance values for each OTU are stacked in order from greatest to least, separate by a thin horizontal line. As long as the parameters you choose to separate the data result in more than one OTU abundance value at the respective position in the plot, the values will be stacked in order as a means of displaying both the sum total value while still representing the individual OTU abundances.


### plot_bar 可视化丰度
默认展示read数目
```{R}
plot_bar(gp.ch)
```

按照OTU所属的属水平进行上色
><font size=2>Add fill color to represent the Genus to which each OTU belongs.

```{R}
plot_bar(gp.ch, fill="Genus")
```


译者补充：X坐标默认显示样品，这里设置使用分组文件

下面fill填充的颜色不变，将x轴替换为分组信息，这个分组信息代表了样本取样来源。

><font size=2>Now keep the same fill color, and group the samples together by the SampleType variable; essentially, the environment from which the sample was taken and sequenced.

```{R}
plot_bar(gp.ch, x="SampleType", fill="Genus")
```

注意我们同一个属的数据都使用相同的颜色映射，因此，相同的颜色可能代表多个OTU，这些OTU之间使用分隔线分开，这些分隔线如果在丰度很低的情况下重叠严重，可能影响对结果的解读。
注意这里柱子的筒高度为全部的丰度总和，这也完全依赖于使用者的标准化方式，议案来讲，我们使用相对丰度标准化之后柱子的总高度都会变得一样，这也是我们常见的可视化方式。

><font size=2>abundance values for the same OTU from the same SampleType will be stacked as separate bar segments, and so the segment lines may not accurately portray the observed richness (because the same OTU might be shown more than once for the same horizontal axis grouping). However, all other aspects of the representation are quantitative, with the total stacked bar height at each horizontal position indicating the sum of all reads for that sample(s). There is not attempt by plot_bar to normalize or standardize your data, which is your job to do (using other tools in the phyloseq pacakge, for instance) before attempting to interpret/compare these values between samples




### 使用分面组织可视化方式
**More Sophisticated Organization using Facets**

使用科水平的物种作为X轴对象，并且使用SampleType信息对作为分面展示不同分组的科水平物种丰度，使用属填充柱状图。这里可以进一步分离数据，使得在每个取样组中更加清楚观察不同属之间的丰度变化。


><font size=2>In the following example we elected to further organize the data using “facets” – separate, adjacent sub-plots. In this case the facets allow us to according to the genus of each OTU. Within each genus facet, the data is further separated by sequencing technology, and the enterotype label for the sample from which each OTU originated is indicated by fill color.


译者补充：facet_grid参数：分组作为分面展示

```{R}
plot_bar(gp.ch, "Family", fill="Genus", facet_grid=~SampleType)
```





### 基于ggplot2的图形修饰 添加点
**Further customization using ggplot2 layers**

ggplot2出图总是很容易被我们修改，就像提供了一个橡皮擦一样，这里我们将丰度映射为抖动的点并添加到图形中。

><font size=2>Note that additional customizations of the plot are always possible using standard ggplot2 layers. For example, the following code chunk shows a plot with jittered points add using a second plot layer.

```{R}
library("ggplot2")
p = plot_bar(gp.ch, "Family", fill="Genus", facet_grid=~SampleType)
p + geom_point(aes(x=Family, y=Abundance), color="black", position="jitter", size=3)

```
### 对Enterotypess数据集进行分析
**Enterotypes dataset examples**&


提取前十丰度的OTU，过滤phyloseq对象

><font size=2>First, load package (if you haven’t already), then trim Enterotype data to most abundant 10 genera.


```{R}
library("phyloseq")
data("enterotype")
TopNOTUs <- names(sort(taxa_sums(enterotype), TRUE)[1:10])
# ent10   <- prune_species(TopNOTUs, enterotype)

ent10   <- prune_taxa(TopNOTUs, enterotype)
ent10
```


```{R}
plot_bar(ent10, "SeqTech", fill="Enterotype", facet_grid=~Genus)
```
我们可以通过变化各种分组来选择更好的可视化方式，我们也建议您在太多自己的数据是尝试不同的参数设置，其实，处理分组文件（sample_data)中的可以被指定为分组之外，plot——bar还识别tax-table中的物种分类名称（如果存在的话）。在上面的这个例子中我们就是使用的属作为分面变量。在每个分面中数据又使用测序技术和不同的肠型进行进一步分开。通过这一分组，我们很清楚的看到测序技术Udine检测到的属的影响，以及该属OTU的数量特征。

><font size=2>The parameters to plot_bar in the following code-chunk were chosen after various trials. We suggest that you also try different parameter settings while you’re exploring different features of the data. In addition to the variables names of sample_data, the plot_bar function recognizes the names of taxonomic ranks, if present. In this example we have also elected to organize data by “facets” (separate, adjacent sub-plots) according to the genus of each OTU. Within each genus facet, the data is further separated by sequencing technology, and the enterotype label for the sample from which each OTU originated is indicated by fill color. Abundance values from different samples and OTUs but having the same variables mapped to the horizontal (x) axis are sorted and stacked, with thin horizontal lines designating the boundaries. With this display it is very clear that the choice of sequencing technology had a large effect on which genera were detected, as well as the fraction of OTUs that were assigned to a Genus.










### 多个分组进行分面

可以同时使用测序平台和不同肠型进行分面，使用属为填充颜色。这只需要做简单的修改即可。

><font size=2>You could nix the approach in which OTU abundance values from different samples, different enterotypes, are stacked together and simply shaded differently, and instead opt to separate both the enterotype designation of the samples and the genus designation of the OTUs into one grid. Only a slight modification to the previous function call is necessary in that case (with an added fill to make it even easier to read):


```{R}
plot_bar(ent10, "Genus", fill="Genus", facet_grid=SeqTech~Enterotype)
```
### 修饰出图
**Add ggplot2 layer to remove the OTU separation lines**

使用ggplot2可视化，很容易进行图形的修改和图层的添加，这里我们添加柱状图。（之前在教程中我们有展示过，这里我们重新进行一个展示）


><font size=2>The following example uses more ggplot2-package commands directly for customization, so you need to load the package first (which we did earlier, but I will show it again here for modularity).

color=Genus将柱状图外边边框使用属进行填充
```{R}
library("ggplot2")
p = plot_bar(ent10, "Genus", fill="Genus", facet_grid=SeqTech~Enterotype)
p + geom_bar(aes(color=Genus, fill=Genus), stat="identity", position="stack")
```


## reference
https://joey711.github.io/phyloseq/plot_bar-examples.html


